src/main/java/org/broadinstitute/hellbender/tools/spark/pathseq/PSScoreUtils.java

```diff
@@ -0,0 +1,426 @@
+     * Gets taxonomic IDs of contigs that aligned sufficiently well to the reads. If both mates are present, returns the
+     * intersection of their hits. Also sets read tag HITS_TAG to a comma-separated list of the IDs.
+     */
+    static JavaRDD<Tuple2<Iterable<GATKRead>, PSPathogenAlignmentHit>> mapGroupedReadsToTax(final JavaRDD<Iterable<GATKRead>> pairs,
+                                                                                            final double minCoverage,
+                                                                                            final double minIdentity,
+                                                                                            final Broadcast<Map<String, String>> refNameToTaxBroadcast) {
+        return pairs.map(readIter -> {
+
+            final Collection<List<String>> taxIDLists = Utils.stream(readIter)
+                    //Flat map the taxID's of all alignments meeting the coverage/identity criteria.
+                    //Throw out duplicates so it returns a list of unique tax ID's for each read in the pair
+                    .flatMap(read -> getValidHits(read, minCoverage, minIdentity).stream()
+                            .map(contig -> refNameToTaxBroadcast.value().containsKey(contig) ? refNameToTaxBroadcast.value().get(contig) : null)
+                            .filter(Objects::nonNull).distinct()
+                    ).collect(Collectors.groupingBy(Function.identity())).values(); //Group the flattened stream by tax ID into lists
+
+            final int numReads = (int) Utils.stream(readIter).count();
+
+            //Count the length of each list - if it's equal to the number of reads, we have a hit
+            final List<String> hitTaxIds = taxIDLists.stream()
+                    .filter(ids -> ids.size() == numReads)
+                    .map(ids -> ids.iterator().next())
+                    .collect(Collectors.toList());
+
+            final PSPathogenAlignmentHit info = new PSPathogenAlignmentHit(hitTaxIds, numReads);
+
+            //If there was at least one hit, append a tag to each read with the list of hits
+            if (hitTaxIds.size() > 0) {
+                final StringBuilder stringBuilder = new StringBuilder();
+                for (final String hit : hitTaxIds) {
+                    stringBuilder.append(hit);
+                    stringBuilder.append(",");
+                }
+                final String hitString = stringBuilder.substring(0, stringBuilder.length() - 1);
+                Utils.stream(readIter).forEach(read -> read.setAttribute(HITS_TAG, hitString));
+            } else {
+                Utils.stream(readIter).forEach(GATKRead::setIsUnmapped);
+            }
+            return new Tuple2<>(readIter, info);
+        });
+    }
+
+
+    /**
+     * Gets set of sufficiently well-mapped hits
+     */
+    private static Collection<String> getValidHits(final GATKRead read, final double minCoverage, final double minIdentity) {
```

Comment:
I prefer to use a `List` over a `Collection` for the return type of something like this. That way at least the caller knows some semantics about the collection type (ie there might be duplicates)